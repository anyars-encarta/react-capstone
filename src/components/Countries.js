// Countries.js
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCircleRight } from '@fortawesome/free-solid-svg-icons';
import {
  setCountries, selectCountries, setError, setLoading,
} from '../redux/countriesSlice';
import '../styles/Countries.css';

function Countries() {
  const dispatch = useDispatch();
  const countries = useSelector(selectCountries);
  const { continentName } = useParams();
  const [continentStats, setContinentStats] = useState({
    totalCases: 0,
    totalRecoveries: 0,
    totalDeaths: 0,
  });
  const [searchQuery, setSearchQuery] = useState('');

  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await fetch('https://covid19-update-api.herokuapp.com/api/v1/world/country/:country');
        const data = await response.json();
        dispatch(setCountries(data));
        dispatch(setError(null));
        dispatch(setLoading(false));
      } catch (error) {
        dispatch(setError(error.message));
        dispatch(setLoading(false));
      }
    };

    fetchData();
  }, [dispatch]);

  useEffect(() => {
    if (Array.isArray(countries.list)) {
      const continentStatsData = countries.list
        .filter((country) => country.continent === continentName);
      const continentStatsResult = continentStatsData.reduce((stats, country) => ({
        ...stats,
        totalCases: stats.totalCases + parseInt(country.cases, 10),
        totalRecoveries: stats.totalRecoveries + parseInt(country.recovered, 10),
        totalDeaths: stats.totalDeaths + parseInt(country.deaths, 10),
      }), {
        totalCases: 0,
        totalRecoveries: 0,
        totalDeaths: 0,
      });

      setContinentStats(continentStatsResult);
    }
  }, [continentName, countries]);

  const formatNumber = (number) => number.toLocaleString();

  const handleSearchInputChange = (e) => {
    setSearchQuery(e.target.value);
  };

  const filteredCountries = continentName
    ? countries.list.filter(
      (country) => country.continent === continentName
          && country.country.toLowerCase().includes(searchQuery.toLowerCase()),
    )
    : [];

  if (countries.loading) {
    return <div>Loading...</div>;
  }

  if (countries.error) {
    return (
      <div>
        Error:
        {countries.error}
      </div>
    );
  }

  return (
    <>
      <div className="top-stats">
        <div className="top">
          <p className="continental">
            Stats for
            {' '}
            {continentName}
          </p>
          <p className="count">
            Cases:
            {' '}
            {formatNumber(continentStats.totalCases)}
          </p>
          <p className="count">
            Recoveries:
            {' '}
            {formatNumber(continentStats.totalRecoveries)}
          </p>
          <p className="count">
            Deaths:
            {' '}
            {formatNumber(continentStats.totalDeaths)}
          </p>
        </div>
        <div className="search-container">
          <input
            type="text"
            id="text"
            placeholder="Search countries..."
            value={searchQuery}
            onChange={handleSearchInputChange}
          />
          <FontAwesomeIcon icon={faCircleRight} className="arrow-right" />
        </div>
      </div>
      <p className="countries-stats">COUNTRIES STATS</p>
      <div className="country-container">
        {filteredCountries.map((country) => (
          <div key={country.country} className="country">
            <div className="overlay" />
            <div className="info">
              <p className="country-name">{country.country}</p>
              <p className="cases-count">
                Cases:
                {' '}
                {formatNumber(country.cases)}
              </p>
              <p className="recovery-count">
                Recovered:
                {' '}
                {formatNumber(country.recovered)}
              </p>
              <p className="deaths-count">
                Deaths:
                {' '}
                {formatNumber(country.deaths)}
              </p>
              <p className="active-count">
                Active:
                {' '}
                {formatNumber(country.active)}
              </p>
              <p className="today-count">
                Today&apos;s Cases:
                {' '}
                {formatNumber(country.todayCases)}
              </p>
              <p className="today-death-count">
                Today&apos;s Deaths:
                {' '}
                {formatNumber(country.todayDeaths)}
              </p>
              <p className="today-recv-count">
                Today&apos;s Recovery:
                {' '}
                {formatNumber(country.todayRecovered)}
              </p>
              <p className="critical-count">
                Critical:
                {' '}
                {formatNumber(country.critical)}
              </p>
              <p className="tests-count">
                Tests:
                {' '}
                {formatNumber(country.tests)}
              </p>
              <p className="population-count">
                Population:
                {' '}
                {formatNumber(country.population)}
              </p>
            </div>
            <FontAwesomeIcon icon={faCircleRight} className="view-more" />
          </div>
        ))}
      </div>
    </>
  );
}

export default Countries;
