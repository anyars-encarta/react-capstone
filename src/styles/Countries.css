.top-stats .continental {
  margin: 0;
  font-size: 40px;
}

.country-container {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
  background-color: rgba(225, 73, 132, 0.75);
}

.countries-stats {
  margin: 0;
  padding: 2px;
  background-color: #e14984;
  color: #ffff;
  font-weight: bold;
}

.country {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  justify-content: right;
  width: 100%;
  padding: 10px;
  color: white;
  position: relative;
  cursor: pointer;
  margin: 0;
  box-sizing: border-box;
}

.country:nth-child(4n-1),
.country:nth-child(4n-3) {
  background-color: rgba(225, 73, 132); /* Background color for cards 3 and 4 */
}

.country-name {
  font-size: 30px;
}

.search-container {
  display: flex;
  justify-content: space-between;
}

#text {
  border-radius: 50px;
  padding: 5px;
  width: 70%;
  background-color: rgba(225, 73, 132, 0.75);
  border: 2px solid white;
  color: white;
}
