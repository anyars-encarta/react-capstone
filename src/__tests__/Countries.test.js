import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom/extend-expect'; // For expect(...).toBeInTheDocument()
import Countries from '../components/Countries';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('Countries Integration Test', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      countries: {
        list: [
          // Mock your countries data here
          {
            country: 'Country1',
            continent: 'ContinentName',
            cases: 100,
            recovered: 50,
            deaths: 5,
            active: 45,
            todayCases: 10,
            todayDeaths: 2,
            todayRecovered: 8,
            critical: 3,
            tests: 1000,
            population: 1000000,
          },
          {
            country: 'Country2',
            continent: 'ContinentName',
            cases: 200,
            recovered: 100,
            deaths: 10,
            active: 90,
            todayCases: 20,
            todayDeaths: 4,
            todayRecovered: 16,
            critical: 6,
            tests: 2000,
            population: 2000000,
          },
          // Add more mock data as needed
        ],
        loading: false,
        error: null,
        stats: {
          totalCases: 300,
          totalRecoveries: 150,
          totalDeaths: 15,
        },
      },
    });
  });

  it('should render the component with data', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/countries/ContinentName']}>
          <Route path="/countries/:continentName">
            <Countries />
          </Route>
        </MemoryRouter>
      </Provider>,
    );

    // Assert that the component has rendered correctly
    expect(screen.getByText('Stats for ContinentName')).toBeInTheDocument();
    expect(screen.getByText('Cases: 300')).toBeInTheDocument();
    expect(screen.getByText('Recoveries: 150')).toBeInTheDocument();
    expect(screen.getByText('Deaths: 15')).toBeInTheDocument();
    expect(screen.getByText('Country1')).toBeInTheDocument();
    expect(screen.getByText('Country2')).toBeInTheDocument();
  });

  it('should filter countries based on search query', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/countries/ContinentName']}>
          <Route path="/countries/:continentName">
            <Countries />
          </Route>
        </MemoryRouter>
      </Provider>,
    );

    // Enter search query
    const searchInput = screen.getByPlaceholderText('Search countries...');
    fireEvent.change(searchInput, { target: { value: 'Country1' } });

    // Assert that only Country1 is displayed
    expect(screen.getByText('Country1')).toBeInTheDocument();
    expect(screen.queryByText('Country2')).not.toBeInTheDocument();
  });

  // Add more integration tests as needed

  it('should match snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/countries/ContinentName']}>
          <Route path="/countries/:continentName">
            <Countries />
          </Route>
        </MemoryRouter>
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
