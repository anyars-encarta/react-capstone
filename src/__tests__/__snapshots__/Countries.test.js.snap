// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countries Integration Test should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="top-stats"
  >
    <div
      class="top"
    >
      <p
        class="continental"
      >
        Stats for ContinentName
      </p>
      <p
        class="count"
      >
        Cases: 300
      </p>
      <p
        class="count"
      >
        Recoveries: 150
      </p>
      <p
        class="count"
      >
        Deaths: 15
      </p>
    </div>
    <div
      class="search-container"
    >
      <input
        id="text"
        placeholder="Search countries..."
        type="text"
        value=""
      />
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-right arrow-right"
        data-icon="circle-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 256a256 256 0 1 0 512 0A256 256 0 1 0 0 256zM294.6 135.1l99.9 107.1c3.5 3.8 5.5 8.7 5.5 13.8s-2 10.1-5.5 13.8L294.6 376.9c-4.2 4.5-10.1 7.1-16.3 7.1C266 384 256 374 256 361.7l0-57.7-96 0c-17.7 0-32-14.3-32-32l0-32c0-17.7 14.3-32 32-32l96 0 0-57.7c0-12.3 10-22.3 22.3-22.3c6.2 0 12.1 2.6 16.3 7.1z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
  <p
    class="countries-stats"
  >
    COUNTRIES STATS
  </p>
  <div
    class="country-container"
  >
    <div
      class="country"
    >
      <div
        class="overlay"
      />
      <div
        class="info"
      >
        <p
          class="country-name"
        >
          Country1
        </p>
        <p
          class="cases-count"
        >
          Cases: 100
        </p>
        <p
          class="recovery-count"
        >
          Recovered: 50
        </p>
        <p
          class="deaths-count"
        >
          Deaths: 5
        </p>
        <p
          class="active-count"
        >
          Active: 45
        </p>
        <p
          class="today-count"
        >
          Today's Cases: 10
        </p>
        <p
          class="today-death-count"
        >
          Today's Deaths: 2
        </p>
        <p
          class="today-recv-count"
        >
          Today's Recovery: 8
        </p>
        <p
          class="critical-count"
        >
          Critical: 3
        </p>
        <p
          class="tests-count"
        >
          Tests: 1,000
        </p>
        <p
          class="population-count"
        >
          Population: 1,000,000
        </p>
      </div>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-right view-more"
        data-icon="circle-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 256a256 256 0 1 0 512 0A256 256 0 1 0 0 256zM294.6 135.1l99.9 107.1c3.5 3.8 5.5 8.7 5.5 13.8s-2 10.1-5.5 13.8L294.6 376.9c-4.2 4.5-10.1 7.1-16.3 7.1C266 384 256 374 256 361.7l0-57.7-96 0c-17.7 0-32-14.3-32-32l0-32c0-17.7 14.3-32 32-32l96 0 0-57.7c0-12.3 10-22.3 22.3-22.3c6.2 0 12.1 2.6 16.3 7.1z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="country"
    >
      <div
        class="overlay"
      />
      <div
        class="info"
      >
        <p
          class="country-name"
        >
          Country2
        </p>
        <p
          class="cases-count"
        >
          Cases: 200
        </p>
        <p
          class="recovery-count"
        >
          Recovered: 100
        </p>
        <p
          class="deaths-count"
        >
          Deaths: 10
        </p>
        <p
          class="active-count"
        >
          Active: 90
        </p>
        <p
          class="today-count"
        >
          Today's Cases: 20
        </p>
        <p
          class="today-death-count"
        >
          Today's Deaths: 4
        </p>
        <p
          class="today-recv-count"
        >
          Today's Recovery: 16
        </p>
        <p
          class="critical-count"
        >
          Critical: 6
        </p>
        <p
          class="tests-count"
        >
          Tests: 2,000
        </p>
        <p
          class="population-count"
        >
          Population: 2,000,000
        </p>
      </div>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-right view-more"
        data-icon="circle-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 256a256 256 0 1 0 512 0A256 256 0 1 0 0 256zM294.6 135.1l99.9 107.1c3.5 3.8 5.5 8.7 5.5 13.8s-2 10.1-5.5 13.8L294.6 376.9c-4.2 4.5-10.1 7.1-16.3 7.1C266 384 256 374 256 361.7l0-57.7-96 0c-17.7 0-32-14.3-32-32l0-32c0-17.7 14.3-32 32-32l96 0 0-57.7c0-12.3 10-22.3 22.3-22.3c6.2 0 12.1 2.6 16.3 7.1z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
